lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -311,14 +311,26 @@ defmodule DynamicSupervisor do
   @doc since: "1.6.0"
   @spec start_child(Supervisor.supervisor(), Supervisor.child_spec() | {module, term} | module) ::
           on_start_child()
-  def start_child(supervisor, {_, _, _, _, _, _} = child_spec) do
-    validate_and_start_child(supervisor, child_spec)
+  @spec start_child(Supervisor.supervisor(), deprecated_child_spec :: Supervisor.Spec.spec()) ::
+          on_start_child()
```

Comment:
I honestly don't see the benefit of this change. We are duplicating all of the clauses in `Supervisor.child_spec/2` here. What is the gain?

Reply:
Well, they are not deprecated, because their usage in this module hasn't been formally depracted, but `Supervisor.Spec` has. The gain is of having two specs is that you can show in the second one that its usage is discourage (something like `discouraged_child_spec :: Supervisor.Spec.spec()` can be used instead), and user only has to look at the first one.

Reply:
They are not deprecated and they will never be, unless Erlang deprecates them, which is unlikely. Otherwise we can’t use DynamicSupervisor with an Erlang lib that returns old type specs.

Reply:
Can you please send a PR only with the docs changes? The code changes are incorrect per the reason above, thank you.

Reply:
Sorry, but this PR does three things. 1. fixes documentation. 2. fixes the specs which still needs to be fixed regardless of whether they are split into two, or labeled as "discouraged". 3. fixes the code to emit a FunctionClauseError when a spec that is not covered by the specs.
The changes to the code do not change at all the way the code deals when provided a spec in the old format (6-tuple one)

Reply:
The code change is undesired, there is no reason to duplicate the work done by Supervisor.child_spec.

Reply:
Also we should not be referring to Supervisor.Spec. If you need to, mention “old Erlang child specs”.

Reply:
For the spec bit, you can use :supervisor.child_spec instead of Supervisor.child_spec.

Reply:
Ping! Can you please update the PR to only do 1 and 2 of what you described above? Thank you.

Reply:
done